python/dtype.cpp

```diff
@@ -55,3 +61,48 @@ scipp::core::DType scipp_dtype(const py::object &type) {
     return scipp_dtype(py::dtype::from_args(type));
   }
 }
+
+// Awful handwritten parser. But std::regex does not support string_view and
+// you have to go through c-strings in order to extract the unit scale.
+[[nodiscard]] scipp::units::Unit
+parse_datetime_dtype(const std::string_view dtype_name) {
+  if (const auto length = std::size(dtype_name); length == 13 || length == 14) {
+    if (dtype_name.substr(0, 11) == "datetime64[" &&
+        dtype_name[length - 2] == 's' && dtype_name.back() == ']') {
+      if (length == 13) {
+        // no scale given
+        return scipp::units::s;
+      }
+      switch (dtype_name[11]) {
+      case 'n':
+        return scipp::units::ns;
+      case 'u':
+        return scipp::units::us;
+      default:
+        throw std::invalid_argument(
+            std::string("Unsupported unit in datetime: ") + dtype_name[11] +
+            "s");
+      }
+    }
+  }
+
+  throw std::invalid_argument(
+      std::string("Invalid dtype, expected datetime64, got ")
+          .append(dtype_name));
```

Comment:
Have you considered throwing `except::TypeError`? We still have not bound our C++ exceptions to Python exceptions, but once we do that may be more useful?

Reply:
No, I have always regarded the dtype as an argument. But if you see it as inferred from the value-argument, a `TypeError` makes sense?